// xanoscript-lint/lib/validator.js
import fs from 'fs/promises';
import path from 'path';
import { XanoScriptLSPClient } from './lsp-client.js';

/**
 * XanoScript file validator
 * Validates .xs files using the LSP server
 */
export class XanoScriptValidator {
  constructor(serverPath = null) {
    this.client = new XanoScriptLSPClient(serverPath);
    this.results = [];
  }

  /**
   * Start the LSP client
   */
  async start() {
    await this.client.start();
  }

  /**
   * Validate a single file
   */
  async validateFile(filePath) {
    const absolutePath = path.resolve(filePath);

    try {
      const content = await fs.readFile(absolutePath, 'utf8');
      const diagnostics = await this.client.validateFile(absolutePath, content);

      const result = {
        file: filePath,
        absolutePath,
        diagnostics: diagnostics.map(d => this.formatDiagnostic(d)),
        errorCount: diagnostics.filter(d => d.severity === 1).length,
        warningCount: diagnostics.filter(d => d.severity === 2).length,
        infoCount: diagnostics.filter(d => d.severity === 3 || d.severity === 4).length
      };

      this.results.push(result);
      return result;

    } catch (err) {
      const result = {
        file: filePath,
        absolutePath,
        error: err.message,
        diagnostics: [],
        errorCount: 1,
        warningCount: 0,
        infoCount: 0
      };

      this.results.push(result);
      return result;
    }
  }

  /**
   * Validate multiple files
   */
  async validateFiles(filePaths) {
    const results = [];

    for (const filePath of filePaths) {
      const result = await this.validateFile(filePath);
      results.push(result);
    }

    return results;
  }

  /**
   * Format a diagnostic message
   */
  formatDiagnostic(diagnostic) {
    return {
      line: diagnostic.range.start.line + 1,
      column: diagnostic.range.start.character + 1,
      endLine: diagnostic.range.end.line + 1,
      endColumn: diagnostic.range.end.character + 1,
      severity: this.getSeverityString(diagnostic.severity),
      message: diagnostic.message,
      code: diagnostic.code,
      source: diagnostic.source || 'xanoscript'
    };
  }

  /**
   * Convert severity number to string
   */
  getSeverityString(severity) {
    switch (severity) {
      case 1: return 'error';
      case 2: return 'warning';
      case 3: return 'info';
      case 4: return 'hint';
      default: return 'unknown';
    }
  }

  /**
   * Get summary statistics
   */
  getSummary() {
    const totalErrors = this.results.reduce((sum, r) => sum + r.errorCount, 0);
    const totalWarnings = this.results.reduce((sum, r) => sum + r.warningCount, 0);
    const totalInfo = this.results.reduce((sum, r) => sum + r.infoCount, 0);
    const filesWithIssues = this.results.filter(r => r.errorCount > 0 || r.warningCount > 0).length;

    return {
      totalFiles: this.results.length,
      filesWithIssues,
      totalErrors,
      totalWarnings,
      totalInfo,
      hasErrors: totalErrors > 0
    };
  }

  /**
   * Shutdown the LSP client
   */
  async shutdown() {
    await this.client.shutdown();
  }
}
